DataProxy: Populate X-Grafana-Referer header

neilfordyce opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**
Adds a new header X-Grafana-Referer to requests via the datasource proxy, which captures the contents of the the original Referer header, prior to its deletion.  Under normal circumstances (when called from a dashboard), this header will contain a dashboard name.

**Why do we need this feature?**
There are other suggested use cases in https://github.com/grafana/grafana/discussions/38367

Knowing what dashboard kicked off a query is useful for troubleshooting issues with slow running queries.  Some systems which use the proxy data source (like OpenTSDB) don't provide tracing.  This provides a convenient way to correlate expensive requests with the source dashboard via slow query logs.

E.g. OpenTSDB logs the request headers along with query stats https://github.com/OpenTSDB/opentsdb/blob/ab656051b4bb18d2935b8a4f7c77141503bfb597/src/stats/QueryStats.java#L318

In addition, I have another internal service which relies on the referer header being set to provide a way of automatically throttling dashboards which create a large number of slow requests over a short time frame.  I could perhaps use X-Grafana-User for that, but going by dashboard name alone works for anonymous users too.  It's helpful in situations where a dashboard with a number of expensive queries is shared widely and lots of engineers open the same dashboard with different time ranges.

**Who is this feature for?**
System admins who need to understand performance issues with a database called by the proxy.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/discussions/38367

**Special notes for your reviewer**:
Extra context: The Referer header was removed to address https://github.com/grafana/grafana/issues/13949.  That fixed issues for some data sources, but inadvertently caused some issues for services which relied on it.

neilfordyce committed:
ProxyUtil: Populate X-Grafana-Referer header

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=60040) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=60040) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=60040) it.</sub>

sakjur reviewed (approved):
Looks good to me, thanks for adding this!

Leaving this open for a bit to let other maintainers chime in with feedback (i.e. "do we want an option guarding this?" — I don't think that's necessary, but it was my first thought looking at the code) 🙂

marefr commented on pkg/util/proxyutil/reverse_proxy.go:
```diff
@@ -77,6 +77,9 @@ func wrapDirector(d func(*http.Request)) func(req *http.Request) {
 		d(req)
 		PrepareProxyRequest(req)
 
+		// Set X-Grafana-Referer to correlate access logs to dashboards
+		req.Header.Set("X-Grafana-Referer", req.Header.Get("Referer"))
+
```
It would be nice to move this logic and deletion of Origin and Referer headers into  the PrepareProxyRequest function since then it will apply to both datasource proxy and call resource requests:
https://github.com/grafana/grafana/blob/10c2bdea3eec6c143bf59e326eacf000cdbb78d8/pkg/api/plugin_resource.go#L100

neilfordyce replied:
Thanks for the quick feedback.  Good idea.  I moved the logic into PrepareProxyRequest and added a few tests.

marefr reviewed (changes requested):
Make sense. But would like to see a minor change, see comment

neilfordyce committed:
ProxyUtil: Move Referer/Origin header removal

So that the removal and setting X-Grafana-Referer logic applies to all
proxied requests and not just datasource proxy.

marefr reviewed (approved):
LGTM great. Thanks a lot

marefr commented:
Asked @jmatosgrafana to have a look as well in case this have any potential security implications.

jmatosgrafana commented on pkg/util/proxyutil/proxyutil.go:
```diff
@@ -7,9 +7,17 @@ import (
 // Set X-Forwarded-For headers.
 func PrepareProxyRequest(req *http.Request) {
+	// Set X-Grafana-Referer to correlate access logs to dashboards
+	req.Header.Set("X-Grafana-Referer", req.Header.Get("Referer"))
```
As we are re-using as is the Referer value that is an untrusted user input, it could be useful in the unit tests to make sure something like
`Referer: https://www.google.ch\nOtherHeader:https://www.somethingelse.com` does not get processed as 2 different headers: `Referer` and `OtherHeader`

neilfordyce replied:
A test with this would work, but I'm guessing that wouldn't be checking what you're thinking of `req.Header.Set("Referer", "https://www.google.ch\r\nOtherHeader:https://www.somethingelse.com")`.

I could also add something like this, but feels like we'd be testing request parsing more than Grafana.  But then it's effectively another header anyway. 
```
t.Run("Prepare proxy request should set X-Grafana-Referer header as one header", func(t *testing.T) {
  str := "GET / HTTP/1.1\r\n" +
	"Host: blah.com\r\n" +
	"Accept: text/html\r\n" +
	"Referer: https://www.google.ch\r\nOtherHeader:https://www.somethingelse.com\r\n" +
	"Accept-Encoding: gzip, deflate\r\n" +
	"Connection: close\r\n" +
	"\r\n" +
	""
  buf := bufio.NewReader(strings.NewReader(str))
  
  req, err := http.ReadRequest(buf)
  require.NoError(t, err)
  
  PrepareProxyRequest(req)
  require.Contains(t, req.Header, "X-Grafana-Referer")
  require.Equal(t, "https://www.google.ch", req.Header.Get("X-Grafana-Referer"))
})
```

jmatosgrafana replied:
Testing  that `req.Header.Set("Referer", "https://www.google.ch\r\nOtherHeader:https://www.somethingelse.com")` is not producing 2 headers is enough: it "proves" we don't have to care about multiline in this untrusted input and hence we are protected from header injection

neilfordyce replied:
Gotcha.  I added a test to check that in [95d01c6](https://github.com/grafana/grafana/pull/60040/commits/95d01c680e59969e0b2e808b35bc3f056d777fe3)

jmatosgrafana replied:
Thank you. Could we add a `require.NotContains(t, req.Header, "OtherHeader")` to double check ?

neilfordyce replied:
Sure thing, added in [b0a6df9](https://github.com/grafana/grafana/pull/60040/commits/b0a6df9a0a83d51ad55b113e0e8eb028a2cef792)

neilfordyce replied:
I don't think so.  req.Header.Set is essentially just setting a value in a `map[string][]string`.  There's no special encoding/decoding there.   Maybe I'm misunderstanding the risk though.

marefr replied:
Okay just wanted a sanity check

jmatosgrafana reviewed (approved):
LGTM, from a security point of view we may want to double check that a new line in the middle of a Referer value (even if it is unclear how to achieve this) does not result in 2 differents headers being added

neilfordyce committed:
ProxyUtil: Test to guard against multiline headers

neilfordyce committed:
ProxyUtil: Explicitly check injected header isn't parsed

marefr commented:
@neilfordyce thank you for contributing to Grafana
